OCaml_Intro_to_CS/part_8.ml

```diff
@@ -0,0 +1,189 @@
+(*Recall the problem of removing the duplicates from a list.
+Please write another version of remove_duplicates that takes a list of integers and removes
+any duplicates. For example:
+# remove_duplicates [5; 1 ; 1 ; 3 ; 1 ; 5 ; 3 ; 1 ; 5 ; 17];;
+- : int list = [5; 1; 3; 17]
+This time you should take no more than O(n * log n) time and also preserve the order.
+*)
+
+let reverse lst =
+  let rec aux acc = function
+    | [] -> acc
+    | h :: t -> aux (h :: acc) t
+in 
+aux [] lst;;
+
+
+let element_of x hm =
+  let rec aux = function
+    | [] -> false
+    | h :: t -> if h = x then true else aux t
+  in aux hm
+;;
+
+let remove_duplicates lst = 
+  let rec aux acc = function
+    | [] -> acc
+    | h :: t -> if element_of h acc then aux acc t else aux (h :: acc) t
+  in
+  let init = []
+  in
+  reverse (aux init lst)
+;;
```

Comment:
I don't think this is O(n * log(n)), seems like O(n^2) because you search for the element in the accumulator which is O(n). The worst case scenario is that there are no duplicates. You probably need something faster than `element_of`